### ISSUE TITLE
Consider referencing GCP metadata service by IP in Ignition/Afterburn

### ISSUE THREAD

**Issue by bgilbert** [coreos/fedora-coreos-tracker#891]
On GCP, both Ignition and Afterburn access the GCE metadata service via the hostname `metadata.google.internal` without any further authentication of the service, as [instructed](https://cloud.google.com/compute/docs/storing-retrieving-metadata#querying) by official documentation.
An attack that rebinds that hostname can inject arbitrary code (Ignition) or SSH keys (Afterburn).  There have been [multiple](https://gitlab.com/gitlab-com/gl-security/security-operations/gl-redteam/red-team-tech-notes/-/tree/master/oslogin-privesc-june-2020) [exploits](https://github.com/irsl/gcp-dhcp-takeover-code-exec) that involve poisoning that hostname via forged DHCP replies and a vulnerable DHCP hook called `google_set_hostname`.  We don't ship that script, but since DNS poisoning in this case leads to trivial instance takeover, we should perhaps consider mitigations.

`metadata.google.internal` resolves to 169.254.169.254.  One option is to just hardcode that IP.  It's not clear that Google has documented the address as stable, but it'd probably be hard to change it at this point.  Other clouds use the same address.

In the case of Ignition, another potential option is to query the metadata service before a globally routable address has been configured.  However, that would require structural changes to Ignition, since merged configs are fetched in the same Ignition stage as the main config.  It also wouldn't help Afterburn.

xref https://github.com/coreos/fedora-coreos-tracker/issues/885

**Comment by lucab** [comment]
Looking a bit more into GCP related material, the explicit 169.254.169.254 address is already referenced in a few places:
 * in the docs, https://cloud.google.com/compute/docs/internal-dns/#access_by_internal_DNS
 * in their Go client library, https://github.com/googleapis/google-cloud-go/blob/v0.86.0/compute/metadata/metadata.go#L39

**Comment by jlebon** [comment]
```
12:51:03 < bgilbert> #agreed we will switch Ignition and Afterburn to contact the GCP metadata
                     service by IP instead of hostname, unless we receive significant information
                     to the contrary by next week
```

**Comment by bgilbert** [comment]
PRs in https://github.com/coreos/ignition/pull/1247 and https://github.com/coreos/afterburn/pull/595.

**Comment by bgilbert** [comment]
Both PRs have merged upstream.  Closing this out.

**Comment by dustymabe** [comment]
Those changes landed in Ignition `2.12.0` and Afterburn `5.1.0`. Those versions are in the current latest stable.

The fix for this went into stable stream release `35.20211029.3.0`.
